File: app/src/main/java/starbright/com/projectegg/features/detail/RecipeDetailPresenter.java
package starbright.com.projectegg.features.detail;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import starbright.com.projectegg.data.AppRepository;
import starbright.com.projectegg.data.local.model.Recipe;
import starbright.com.projectegg.util.scheduler.BaseSchedulerProvider;

class RecipeDetailPresenter implements RecipeDetailContract.Presenter {

    private final AppRepository mRepository;
    private final RecipeDetailContract.View mView;
    private final BaseSchedulerProvider mSchedulerProvider;
    private CompositeDisposable mCompositeDisposable;

    private String mRecipeId;
    private Recipe mRecipe;

    RecipeDetailPresenter(AppRepository repository,
                          RecipeDetailContract.View view,
                          BaseSchedulerProvider schedulerProvider) {
        mRepository = repository;
        mView = view;
        mView.setPresenter(this);
        mSchedulerProvider = schedulerProvider;
        mCompositeDisposable = new CompositeDisposable();
    }

    @Override
    public void start() {
        mView.setupAdView();
        mView.setupSwipeRefreshLayout();
        getRecipeDetailInformation(mRecipeId);
    }

    @Override
    public void setRecipeId(String recipeId) {
        mRecipeId = recipeId;
    }

    @Override
    public void getRecipeDetailInformation(String recipeId) {
        mView.hideScrollContainer();
        mView.showProgressBar();
        mCompositeDisposable
                .add(mRepository.getRecipeDetailInformation(recipeId)
                        .subscribeOn(mSchedulerProvider.computation())
                        .observeOn(mSchedulerProvider.ui())
                        .subscribe(new Consumer<Recipe>() {
                            @Override
                            public void accept(Recipe recipe) {
                                mView.hideProgressBar();
                                mRecipe = recipe;
                                if (mRecipe != null) {
                                    mView.hideEmptyStateTextView();
                                    updateView();
                                } else {
                                    mView.renderEmptyStateTextView();
                                }
                            }
                        }, new Consumer<Throwable>() {
                            @Override
                            public void accept(Throwable throwable) {
                                mView.hideProgressBar();
                                if (mRecipe != null) {
                                    mView.hideEmptyStateTextView();
                                    updateView();
                                } else {
                                    mView.renderEmptyStateTextView();
                                }
//                                mView.renderErrorStateTextView(throwable.getMessage());
                            }
                        }));
    }

    @Override
    public void handleShareMenuClicked() {
        mView.createShareIntent(mRecipe.getSourceStringUrl(), mRecipe.getTitle());
    }

    @Override
    public void handleWebViewMenuClicked() {
        mView.navigateToWebViewActivity(mRecipe.getSourceStringUrl());
    }

    private void updateView() {
        mView.showScrollContainer();
        mView.renderBannerFoodImage(mRecipe.getImage());
        mView.renderHeaderContainer(mRecipe.getServingCount(), mRecipe.getPreparationMinutes(),
                mRecipe.getCookingMinutes(), mRecipe.getTitle());
        mView.renderIngredientCard(mRecipe.getIngredients());
        mView.renderInstructionCard(mRecipe.getInstructions());
    }

    private boolean isIngredientsEmpty() {
        return mRecipe.getIngredients() == null || mRecipe.getIngredients().isEmpty();
    }

    private boolean isInstructionEmpty() {
        return mRecipe.getInstructions() == null || mRecipe.getInstructions().isEmpty();
    }
}

File: app/src/main/java/starbright/com/projectegg/features/detail/RecipeDetailPresenter.kt
/*
 * Copyright (c) by Andreas (oentoro.andreas@gmail.com)
 * created at 31 - 7 - 2020.
 */

package starbright.com.projectegg.features.detail

import io.reactivex.disposables.CompositeDisposable
import starbright.com.projectegg.R
import starbright.com.projectegg.data.AppRepository
import starbright.com.projectegg.data.model.Recipe
import starbright.com.projectegg.features.base.BasePresenter
import starbright.com.projectegg.util.NetworkHelper
import starbright.com.projectegg.util.scheduler.SchedulerProviderContract
import javax.inject.Inject

class RecipeDetailPresenter @Inject constructor(
    schedulerProvider: SchedulerProviderContract,
    compositeDisposable: CompositeDisposable,
    networkHelper: NetworkHelper,
    private val repository: AppRepository
) : BasePresenter<RecipeDetailContract.View>(schedulerProvider, compositeDisposable, networkHelper), RecipeDetailContract.Presenter {

    private var isBookmarked: Boolean = false
    private var recipe: Recipe? = null

    override fun onCreateScreen() {
        view.setupSwipeRefreshLayout()
    }

    override fun getRecipeDetailInformation(recipeId: String) {
        isRecipeBookmarked(recipeId)
        if (!isConnectedToInternet()) view.showError(R.string.server_connection_error)

        view.run {
            hideScrollContainer()
            showProgressBar()
        }
        compositeDisposable.add(
            repository.getRecipeDetailInformation(recipeId)
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe({ recipe ->
                    view.hideProgressBar()
                    this.recipe = recipe
                    if (this.recipe != null) {
                        view.hideEmptyView()
                        updateView()
                    } else {
                        view.renderEmptyView()
                    }
                }, {
                    view.hideProgressBar()
                    if (recipe != null) {
                        view.hideEmptyView()
                        updateView()
                    } else {
                        view.renderEmptyView()
                    }
                })
        )
    }

    override fun handleShareMenuClicked() {
        recipe?.let {
            val sourceUrl = it.sourceStringUrl
            if (sourceUrl != null) {
                view.createShareIntent(sourceUrl, it.title)
            } else {
                view.showError(R.string.detail_empty_label)
            }
        } ?: view.showError(R.string.detail_empty_label)
    }

    override fun handleWebViewMenuClicked() {
        recipe?.sourceStringUrl?.let {
            view.navigateToWebViewActivity(it)
        } ?: view.showError(R.string.detail_empty_label)
    }

    override fun handleBookmarkRecipeMenuClicked() {
        if (isBookmarked) {
            removeRecipeBookmark()
        } else {
            bookmarkRecipe()
        }
    }

    private fun updateView() {
        recipe?.let { recipe ->
            with(view) {
                showScrollContainer()
                renderBannerFoodImage(recipe.image ?: "")
                renderHeaderContainer(
                    recipe.servingCount ?: 0,
                    recipe.cookingMinutes ?: 0,
                    recipe.title,
                    if (recipe.dishTypes?.isEmpty() == true) "" else recipe.dishTypes?.first()
                        ?: "",
                    recipe.calories ?: 0
                )
                recipe.ingredients?.let { ingredients ->
                    renderIngredientsList(ingredients.toMutableList())
                }
                recipe.instructions?.let { instructions ->
                    renderInstructionsList(instructions.toMutableList())
                }
            }
        }
    }

    private fun bookmarkRecipe() {
        recipe?.let {
            compositeDisposable.add(
                repository.saveFavouriteRecipe(it)
                    .subscribeOn(schedulerProvider.io())
                    .observeOn(schedulerProvider.ui())
                    .subscribe {
                        isBookmarked = true
                        view.showSnackbar(R.string.detail_message_favourite_add)
                        view.updateMenu(isBookmarked)
                    }
            )
        }
    }

    private fun removeRecipeBookmark() {
        recipe?.let {
            compositeDisposable.add(
                repository.removeFavouriteRecipe(it.id)
                    .subscribeOn(schedulerProvider.io())
                    .observeOn(schedulerProvider.ui())
                    .subscribe {
                        isBookmarked = false
                        view.showSnackbar(R.string.detail_message_favourite_remove)
                        view.updateMenu(isBookmarked)
                    }
            )
        }
    }

    private fun isRecipeBookmarked(recipeId: String) {
        compositeDisposable.add(
            repository.isRecipeSavedBefore(recipeId.toInt())
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe {
                    isBookmarked = it != null
                    view.updateMenu(isBookmarked)
                }
        )
    }
}
